controllers/nodemodulesconfig_reconciler.go

```diff
@@ -0,0 +1,681 @@
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/yaml"
+)
+
+type WorkerAction string
+
+const (
+	WorkerActionLoad   = "Load"
+	WorkerActionUnload = "Unload"
+
+	NodeModulesConfigReconcilerName = "NodeModulesConfig"
+
+	actionLabelKey             = "kmm.node.kubernetes.io/worker-action"
+	configAnnotationKey        = "kmm.node.kubernetes.io/worker-config"
+	nodeModulesConfigFinalizer = "kmm.node.kubernetes.io/nodemodulesconfig-reconciler"
+	volumeNameConfig           = "config"
+	workerContainerName        = "worker"
+)
+
+//+kubebuilder:rbac:groups=kmm.sigs.x-k8s.io,resources=nodemodulesconfigs,verbs=get;list;watch
+//+kubebuilder:rbac:groups=kmm.sigs.x-k8s.io,resources=nodemodulesconfigs/status,verbs=patch
+//+kubebuilder:rbac:groups="core",resources=pods,verbs=create;delete;get;list;watch
+//+kubebuilder:rbac:groups="core",resources=nodes,verbs=get;list;watch
+
+type NodeModulesConfigReconciler struct {
+	client client.Client
+	helper WorkerHelper
+}
+
+func NewNodeModulesConfigReconciler(client client.Client, helper WorkerHelper) *NodeModulesConfigReconciler {
+	return &NodeModulesConfigReconciler{
+		client: client,
+		helper: helper,
+	}
+}
+
+func (r *NodeModulesConfigReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	logger := ctrl.LoggerFrom(ctx)
+
+	nmc := kmmv1beta1.NodeModulesConfig{}
+
+	if err := r.client.Get(ctx, req.NamespacedName, &nmc); err != nil {
+		if k8serrors.IsNotFound(err) {
+			// Pods are owned by the NMC, so the GC will have deleted them already.
+			// Remove the finalizer if we did not have a chance to do it before NMC deletion.
+			logger.Info("Clearing worker Pod finalizers")
+
+			if err = r.helper.RemoveOrphanFinalizers(ctx, req.Name); err != nil {
+				return reconcile.Result{}, fmt.Errorf("could not clear all Pod finalizers for NMC %s: %v", req.Name, err)
+			}
+
+			return reconcile.Result{}, nil
+		}
+
+		return reconcile.Result{}, fmt.Errorf("could not get NodeModuleState %s: %v", req.NamespacedName, err)
+	}
+
+	if err := r.helper.SyncStatus(ctx, &nmc); err != nil {
+		return reconcile.Result{}, fmt.Errorf("could not reconcile status for NodeModulesConfig %s: %v", nmc.Name, err)
+	}
+
+	// Statuses are now up-to-date.
+
+	statusMap := make(map[string]*kmmv1beta1.NodeModuleStatus, len(nmc.Status.Modules))
+
+	for i := 0; i < len(nmc.Status.Modules); i++ {
+		status := nmc.Status.Modules[i]
+		statusMap[status.Namespace+"/"+status.Name] = &nmc.Status.Modules[i]
+	}
+
+	for _, mod := range nmc.Spec.Modules {
+		moduleNameKey := mod.Namespace + "/" + mod.Name
+
+		logger := logger.WithValues("module", moduleNameKey)
+
+		if err := r.helper.ProcessModuleSpec(ctrl.LoggerInto(ctx, logger), &nmc, &mod, statusMap[moduleNameKey]); err != nil {
+			return ctrl.Result{}, fmt.Errorf("error while processing module %s: %v", moduleNameKey, err)
```

Comment:
do we need to return, or can we just log and store the error and continue to next module?

Reply:
We can probably store the error and continue.

Reply:
Done.